sql/core/src/main/scala/org/apache/spark/sql/execution/bucketing/DisableUnnecessaryBucketedScan.scala

```diff
@@ -120,7 +120,7 @@ object DisableUnnecessaryBucketedScan extends Rule[SparkPlan] {
 
   private def hasInterestingPartition(plan: SparkPlan): Boolean = {
     plan.requiredChildDistribution.exists {
-      case _: ClusteredDistribution | _: HashClusteredDistribution => true
+      case _: ClusteredDistribution | _: HashClusteredDistribution | AllTuples => true
```

Comment:
`AllTuples` is only interesting when the underlying file scan operator has only 1 bucket. So I think we need have a way to check the scan operator to have only 1 bucket as well.

Reply:
`AllTuples` require a single partition. I don't think the additional check for the number of buckets is required.

Reply:
For a query like below

```
SELECT SUM(id)
FROM t1
```

If table `t1` has more than 1 bucket (say 4 buckets), we can still disable bucket scan, because scan will have 4 RDD partitions/tasks with bucketed scan, and it does not satisfy single partition requirement from `AllTuples`.

Reply:
If you disable bucket scan, then it will not enter the `disableBucketWithInterestingPartition` right?

Reply:
Actually after a second thought, I think the change here should work. For the query like https://github.com/apache/spark/pull/33711#discussion_r687173621, the query plan when entering this rule, would be:

```
HashAggregate(keys=[], functions=[sum(id#41)], output=[sum(id)#45L])
+- Exchange SinglePartition, ENSURE_REQUIREMENTS, [id=#66]
   +- HashAggregate(keys=[], functions=[partial_sum(id#41)], output=[sum#48L])
       +- FileScan
```

So the bucket scan will be automatically disabled by the rule, because there's a shuffle between `HashAggregate` and `FileScan`.

Reply:
It's flaky to fix this issue in an optimizer rule since this rule can be disabled sometime. Is it better that change the outpt partitioning in `FileSourceScanExec` to `UnknownPartitioning` if the bucket is 1 ?

Reply:
This is a bug of the `DisableUnnecessaryBucketedScan` rule, IIUC

Reply:
@ulysses-you - this is a bug in `DisableUnnecessaryBucketedScan` itself. In addition:

> Is it better that change the outpt partitioning in FileSourceScanExec to UnknownPartitioning if the bucket is 1 ?

This will cause regression for some queries, e.g. the aggregation query in https://github.com/apache/spark/pull/33711#discussion_r687173621. An extra shuffle is needed before aggregation, if the output partitioning is changed from `HashPartitioning(id, 1)` to `UnknownPartitioning`.

Reply:
ah I see it, thank you @cloud-fan @c21 !